src/relax/transform/eliminate_common_subexpr.cc

```diff
@@ -20,223 +20,162 @@
+  void VisitBinding(const Binding& binding) override {
+    Expr bound_value = VisitExpr(GetBoundValue(binding));
+
+    Binding output_binding = [&]() -> Binding {
+      if (binding.as<VarBindingNode>()) {
+        return VarBinding(binding->var, bound_value);
+      } else if (auto match_cast = binding.as<MatchCastNode>()) {
+        return MatchCast(binding->var, bound_value, match_cast->struct_info);
+      } else {
+        LOG(FATAL) << "Binding must be either VarBinding or MatchCast, "
+                   << "but was " << binding->GetTypeKey();
       }
-    }
+    }();
 
-    // Only visit the interior of objects that we might still keep
-    // around.  Otherwise, double-counting these would lead to extra
-    // variable bindings.
-    //
-    // Before:
-    //     y = f(a+b)
-    //     z = f(a+b)
-    //
-    // Expected:
-    //     y = f(a+b)  // De-duped from (y==z)
-    //     z = y
-    //
-    // Erroneous output:
-    //     c = a+b    // Incorrect, a+b only has a single usage.
-    //     y = f(c)   // De-duped from
-    //     z = y
-    //
-    if (auto it = count_map_.find(e); it == count_map_.end() || it->second < 2) {
-      ExprVisitor::VisitExpr(e);
-    }
-  }
+    ReplacementKey lookup_key(output_binding);
 
-  // do not visit inner functions: we will do CSE within those
-  void VisitExpr_(const FunctionNode* func) override {}
+    if (call_only_ && !bound_value->IsInstance<relax::CallNode>()) {
+      VLOG(1) << "Since call_only_ is true, it is forbidden to de-duplicate " << bound_value;
 
-  // we are not going to do replacements inside struct info to avoid binding lots of reused shapes
-  void VisitExprDepStructInfoField(const StructInfo& struct_info) override {}
+    } else if (ContainsImpureCall(bound_value)) {
+      VLOG(1) << "Since the expression is impure, cannot de-duplicate " << bound_value;
 
- private:
-  std::unordered_map<Expr, int, StructuralHash, StructuralEqual> count_map_;
-  ImpurityDetector impurity_detector_;
-};
+    } else if (auto it = expr_replacements_.find(lookup_key); it != expr_replacements_.end()) {
+      VLOG(1) << "Value " << bound_value << " has previously been bound as " << it->second
+              << ".  The duplicate binding of this value to " << binding->var
+              << " will be replaced with a trivial binding, "
+              << "and occurrences of " << binding->var << " will be replaced with " << it->second;
+      output_binding = VarBinding(binding->var, it->second);
+      var_remap_.insert({binding->var->vid, it->second});
 
-class CommonSubexprEliminator : public ExprMutator {
- public:
-  explicit CommonSubexprEliminator(
-      std::unordered_map<Expr, int, StructuralHash, StructuralEqual> count_map,
-      bool call_only = false)
-      : count_map_(std::move(count_map)), call_only_(call_only) {}
+    } else {
+      VLOG(1) << "Value " << bound_value << " is bound to " << binding->var
+              << " and may be de-duplicated if it occurs again.";
```

Comment:
I think so, since any future debugging would be useful.  They are disabled by default, and require both a build flag `-DTVM_LOG_DEBUG=1` and a runtime environment variable `TVM_LOG_DEBUG=eliminate_common_subexpr.cc=1` to enable.

(Trying to get in the habit of leaving `VLOG(1)` statements around after longer debugging sessions, rather than just removing the `LOG(DEBUG)` statements.)